idseq_dag/engine/pipeline_step.py

```diff
@@ -113,6 +116,54 @@ def wait_for_input_files(self):
                         time.sleep(5)
             self.input_files_local.append(flist)
 
+    def validate_input_files(self):
+        ''' Validate input files before running the step '''
+        errors = self.get_input_file_validation_errors()
+        if errors and "error_type" in errors and errors["error_type"] != InputErrorType.NO_ERROR:
+            log.write("Invalid input detected for step %s" % self.name)
+            self.write_input_errors_json({
+                "step": self.name,
+                "errors": errors["errors"] if "errors" in errors else [],
+                "error_type": errors["error_type"].name
+            })
+            raise RuntimeError("input files for step %s were not valid" % self.name)
+
+    def get_input_file_validation_errors(self):
+        ''' Return any input file validation errors'''
+        # Steps should overwrite this method if necessary.
+        return {
+            # Step-specific error messages.
+            "errors": [],
+            # The broad type of error that occurred.
+            "error_type": InputErrorType.NO_ERROR
+        }
+
+    def write_input_errors_json(self, error_json):
+        ''' Write an input_errors.json file to the output_dir_s3 for this step, which can be detected by other services like idseq-web. '''
+        log.write("Writing input_errors.json for step %s" % self.name)
+        input_errors_file_basename = "input_errors.json"
+        local_input_errors_file = "%s/%s" % (self.output_dir_local, input_errors_file_basename)
+        s3_input_errors_file = "%s/%s" % (self.output_dir_s3, input_errors_file_basename)
+
+        with open(local_input_errors_file, 'w') as input_errors_file:
+            json.dump(error_json, input_errors_file)
+
+        idseq_dag.util.s3.upload_with_retries(local_input_errors_file, s3_input_errors_file)
+
+    @staticmethod
+    def validate_input_files_min_reads(input_files, min_reads):
+        '''
+            Checks whether input files (fa, fasta, fq, fastq) have the minimum number of reads.
+            Useful helper method that pipeline steps can use for validation.
+        '''
+        errors = []
+
+        for index, input_file in enumerate(input_files):
+            num_reads = count.reads(input_file)
```

Comment:
I think this would have performance implications for large samples. I don't think you need to count lines of the entire file. Can you iterate over the lines one by one and stop once you pass the min_reads threshold?

Reply:
Turns out I can pass a `max_reads` parameter to `count.reads`.

It runs `cat input_file | head -n {max_reads} | wc -l` instead of `cat input_file | wc -l`. Verified in dagtest that this is much faster.

Thanks for catching.

Reply:
Actually, as we just discussed offline, please use the counts file from the previous step to do the validation -- no reason to count twice